modules/packages/HDF5_HL.chpl

```diff
@@ -3516,4 +3517,128 @@ module HDF5_HL {
+                          filenameStart: string, type eltType, param rank) {
+      use BlockDist, HDF5_WAR;
+
+      var filenames: [1..0] string;
+      for f in findfiles(dirname) {
+        if f.startsWith(dirname + '/' + filenameStart:string) &&
+           f.endsWith(".h5") {
+          filenames.push_back(f);
+        }
+      }
+      const Space = {1..filenames.size};
+      const BlockSpace = Space dmapped Block(Space, locs, 1);
+      var files: [BlockSpace] ArrRec(eltType, rank);
+      forall (f, name) in zip(files, filenames) {
+        var locName = name; // copy this string to be local
+        var file_id = H5Fopen(locName.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);
+        var dims: [0..#rank] hsize_t;
+        H5LTget_dataset_info_WAR(file_id, dsetName.c_str(), c_ptrTo(dims), nil, nil);
+        var data: [0..# (* reduce dims)] eltType;
+        readHDF5Dataset(file_id, dsetName, data);
+
+        var rngTup: rank*range;
+
+        for i in 0..#rank {
+          rngTup[i+1] = 1..dims[i]:int;
+        }
+
+        const D = {(...rngTup)};
+
+        f = new ArrRec(data.eltType, rank, D, reshape(data, D));
+        H5Fclose(file_id);
+      }
+      return files;
+    }
+
+    /* Read the dataset named `dsetName` out of the open file that `file_id`
+       refers to.  Store the input into the array `data`.
+       Can read data of type int/uint (size 8, 16, 32, 64), real (size 32, 64),
+       and c_string.
+     */
+    proc readHDF5Dataset(file_id, dsetName: string, data) {
+      type eltType = data.eltType;
+
+      proc getHDF5Type(type eltType) {
+        var hdf5Type: hid_t;
+        select eltType {
+          when int(8) {
+            hdf5Type = H5T_STD_I8LE;
+          }
+          when int(16) {
+            hdf5Type = H5T_STD_I16LE;
+          }
+          when int(32) {
+            hdf5Type = H5T_STD_I32LE;
+          }
+          when int(64) {
+            hdf5Type = H5T_STD_I64LE;
+          }
+          when uint(8) {
+            hdf5Type = H5T_STD_U8LE;
+          }
+          when uint(16) {
+            hdf5Type = H5T_STD_U16LE;
+          }
+          when uint(32) {
+            hdf5Type = H5T_STD_U32LE;
+          }
+          when uint(64) {
+            hdf5Type = H5T_STD_U64LE;
+          }
+          when real(32) {
+            hdf5Type = H5T_IEEE_F32LE;
+          }
+          when real(64) {
+            hdf5Type = H5T_IEEE_F64LE;
+          }
+          when c_string {
+            hdf5Type = H5Tcopy(H5T_C_S1);
+            H5Tset_size(hdf5Type, H5T_VARIABLE);
+          }
+          otherwise {
+            halt("Unhandled type in getHDF5Type: ", eltType:string);
+          }
+        }
+        return hdf5Type;
+      }
+
+      const hdf5Type = getHDF5Type(eltType);
+      H5LTread_dataset(file_id, dsetName.c_str(), hdf5Type, c_ptrTo(data));
+    }
+
+    record ArrRec {
+      type eltType;
+      param rank: int;
+      var D: domain(rank);
+      var A: [D] eltType;
+    }
```

Comment:
What would it take to make it an "official" type for use by Chapel users? I wanted it myself the other day.

Reply:
Also, if we reference it in the above chpldoc comment, we could either copy it into the comment or add chpldoc on it. If desired, rename it to `ArrayWrapper`.

Reply:
> What would it take to make it an "official" type for use by Chapel users? I wanted it myself the other day.

We'd have to come up with a name for it that people like, decide where to put it (maybe ChapelArray?), and document it.  I'd be fine with the name `ArrayWrapper`.

I'll change the name and add/modify some chpldoc comments.